src/cmap/handshake/client_metadata.ts

```diff
@@ -0,0 +1,227 @@
+import * as os from 'os';
+import * as process from 'process';
+
+import { BSON, Int32 } from '../../bson';
+import { MongoInvalidArgumentError } from '../../error';
+import type { MongoOptions } from '../../mongo_client';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const NODE_DRIVER_VERSION = require('../../../package.json').version;
+
+/**
+ * @public
+ * @see https://github.com/mongodb/specifications/blob/master/source/mongodb-handshake/handshake.rst#hello-command
+ */
+export interface ClientMetadata {
+  driver: {
+    name: string;
+    version: string;
+  };
+  os: {
+    type: string;
+    name?: NodeJS.Platform;
+    architecture?: string;
+    version?: string;
+  };
+  platform: string;
+  application?: {
+    name: string;
+  };
+  /** FaaS environment information */
+  env?: {
+    name: 'aws.lambda' | 'gcp.func' | 'azure.func' | 'vercel';
+    timeout_sec?: Int32;
+    memory_mb?: Int32;
+    region?: string;
+    url?: string;
+  };
+}
+
+/** @public */
+export interface ClientMetadataOptions {
+  driverInfo?: {
+    name?: string;
+    version?: string;
+    platform?: string;
+  };
+  appName?: string;
+}
+
+/** @internal */
+export class LimitedSizeDocument {
+  private document = new Map();
+  constructor(private maxSize: number) {}
+
+  private getBsonByteLength() {
+    return BSON.serialize(this.document).byteLength;
+  }
+
+  /** Only adds key/value if the bsonByteLength is less than MAX_SIZE */
+  public ifFitsSets(key: string, value: Record<string, any> | string): boolean {
```

Comment:
Maybe a little perf nit here. Instead of serialising the entire document with each attempted property addition, can we just instead serialise the field being added? I'm proposing we save the current serialised document size as a property, then all we need to do on attempted addition is serialise the new property to get the proposed size being added to the existing doc. Then it's just a check of `propertySize + this.documentSize > this.maxSize`. But since we're never dealing with over 512 bytes maybe it doesn't matter, so feel free to ignore. Just thinking about cutting a little more time off.

Reply:
Yea I agree, and that's easy enough to manage.